Linux-PAM-1.1.2/modules/pam_rootok/pam_rootok.8
.\"     Title: pam_rootok
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 07/08/2010
.\"    Manual: Linux-PAM Manual
.\"    Source: Linux-PAM Manual
.\"  Language: English
.\"
.TH "PAM_ROOTOK" "8" "07/08/2010" "Linux-PAM Manual" "Linux\-PAM Manual"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
pam_rootok \- Gain only root access
.SH "Synopsis"
.fam C
.HP \w'\fBpam_rootok\&.so\fR\ 'u
\fBpam_rootok\&.so\fR [debug]
.fam
.SH "DESCRIPTION"
.PP
pam_rootok is a PAM module that authenticates the user if their
\fIUID\fR
is
\fI0\fR\&. Applications that are created setuid\-root generally retain the
\fIUID\fR
of the user but run with the authority of an enhanced effective\-UID\&. It is the real
\fIUID\fR
that is checked\&.
.SH "OPTIONS"
.PP
\fBdebug\fR
.RS 4
Print debug information\&.
.RE
.SH "MODULE TYPES PROVIDED"
.PP
The
\fBauth\fR,
\fBacct\fR
and
\fBpassword\fR
module types are provided\&.
.SH "RETURN VALUES"
.PP
PAM_SUCCESS
.RS 4
The
\fIUID\fR
is
\fI0\fR\&.
.RE
.PP
PAM_AUTH_ERR
.RS 4
The
\fIUID\fR
is
\fBnot\fR
\fI0\fR\&.
.RE
.SH "EXAMPLES"
.PP
In the case of the
\fBsu\fR(1)
application the historical usage is to permit the superuser to adopt the identity of a lesser user without the use of a password\&. To obtain this behavior with PAM the following pair of lines are needed for the corresponding entry in the
\FC/etc/pam\&.d/su\F[]
configuration file:
.sp
.if n \{\
.RS 4
.\}
.fam C
.ps -1
.nf
.if t \{\
.sp -1
.\}
.BB lightgray adjust-for-leading-newline
.sp -1

# su authentication\&. Root is granted access by default\&.
auth  sufficient   pam_rootok\&.so
auth  required     pam_unix\&.so
      
.EB lightgray adjust-for-leading-newline
.if t \{\
.sp 1
.\}
.fi
.fam
.ps +1
.if n \{\
.RE
.\}
.sp
.SH "SEE ALSO"
.PP

\fBsu\fR(1),
\fBpam.conf\fR(5),
\fBpam.d\fR(5),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_rootok was written by Andrew G\&. Morgan, <morgan@kernel\&.org>\&.


grep-2.6.3/ChangeLog
2010-04-02  Jim Meyering  <meyering@redhat.com>

	version 2.6.3
	* NEWS: Record release date.

	grep: avoid used-undefined error with truncated multibyte input
	* src/dfa.c (addtok_wc): Don't use buf[0] (it's undefined) when
	wcrtomb returns <= 0.

	MBS_SUPPORT-removal: * src/dfa.c (dfastate):

2010-04-01  Jim Meyering  <meyering@redhat.com>

	maint: avoid unnecessary 2nd getenv("TERM")
	* src/main.c (main): Don't call getenv("TERM") twice -- in the same
	expression, even.

	tests: remove all unportable uses of echo
	* src/main.c: Use printf rather than echo -ne in a comment.
	* tests/fedora: Use printf (not echo) also in ok/fail functions.
	* cfg.mk (sc_prohibit_echo_minus_en): New rule, to prohibit
	any future introduction.

	tests: add explicit requirement for en_US.UTF-8
	* tests/char-class-multibyte: Use require_en_utf8_locale_,
	rather than open-coding it.
	* tests/prefix-of-multibyte: Require the locale explicitly.
	* tests/fgrep-infloop: Likewise.
	This fixes test failures that would arise on systems without
	that particular locale.  Reported by Ludovic Courtès.

	tests: new function, to require an en_US UTF8 locale
	* tests/init.cfg (require_en_utf8_locale_): New function.

	tests: use printf, not echo -n, echo -e, or any combination
	* tests/fedora: Using printf is more portable.

	grep: remove unnecessary code
	* src/main.c (print_line_middle): Now that we use RE_ICASE
	(enabled in commit 70e23616, "dfa: rewrite handling of multibyte
	case_fold lexing"), this case-conversion code is useless and wasteful.
	Remove it.

	doc: fix typo: s/AM_V_AT/AM_V_at/
	* doc/Makefile.am (egrep.1 fgrep.1): The former has case consistent
	with its sister variable, AM_V_GEN, but the latter is the one that
	actually works.

	doc: generated files are best made read-only, ...
	...to minimize risk of accidentally modifying the generated file
	rather than its template.  These are tiny, so no risk, but it's
	a good to be consistent, so generated files are easier to spot.
	* doc/Makefile.am (egrep.1 fgrep.1): When generating these files,
	ensure that they too are created read-only.

	doc: generate grep.1 from template
	* doc/Makefile.am (grep.1): New rule.
	(CLEANFILES): Add grep.1 to the list.
	* .gitignore: Add /doc/grep.1
	* doc/grep.in.1: Replace hard-coded "2.5.1-cvs" with @VERSION@.
	Update copyright year list.
	Omit the line-splitting \(co directive so that update-copyright
	will perform future updates automatically.
	Egmont Koblinger reported the outdated version string
	and copyright year list in the man page:
	http://savannah.gnu.org/bugs/?29390

	doc: prepare to generate grep.1
	* doc/grep.1: Rename to...
	* doc/grep.in.1: ...this.

2010-03-31  Eric Blake  <eblake@redhat.com>

	build: avoid another warning
	Noticed on cygwin:
	get-mb-cur-max.c: In function 'main':
	get-mb-cur-max.c:27: error: unused parameter 'argc' [-Wunused-parameter]

	* tests/get-mb-cur-max.c (main): Use argc.

2010-03-31  Paolo Bonzini  <bonzini@gnu.org>

	tests: fix on systems with broken sh
	* tests/Makefile.am (TESTS_ENVIRONMENT): Adjust coreutils remnants.
	* tests/bre.sh: Invoke script with $SHELL if defined.
	* tests/ere.sh: Likewise.
	* tests/spencer1-locale: Likewise.
	* tests/spencer1.sh: Likewise.

	tests: improve empty test
	* tests/empty: Add more tests, note expected failure.

	tests: improve empty test with respect to locales
	* tests/empty: Add tests for multiple locales.

	grep: fix grep -F against empty string
	* src/searchutils.c (is_mb_middle): Do not return true for empty matches
	when p == buf.

	tests: rename empty.sh to empty
	* tests/empty.sh: Rename to...
	* tests/empty: ... this.
	* tests/Makefile.am (TESTS): Adjust.

	tests: convert empty.sh to new style
	* tests/empty.sh: Convert to init.sh, add 10-second timeout.

	tests: use get-mb-cur-max in char-class-multibyte
	* tests/char-class-multibyte: Use get-mb-cur-max to detect UTF-8 support.
	Rewrite previous locale detection code as a grep test.

	tests: fix -Wformat failure
	* tests/get-mb-cur-max (main): Cast MB_CUR_MAX to int.

2010-03-31  Jim Meyering  <meyering@redhat.com>

	doc: add a "Reply-To" to the suggested announcement mail header
	* README-release: Add "Reply-To" with the list address,
	to minimize risk of replies to the other announcement recipients.
	Suggestion from Eric Blake.

2010-03-30  Jim Meyering  <meyering@redhat.com>

	build: avoid compiler warning when building test program
	* tests/Makefile.am (AM_CPPFLAGS, AM_CFLAGS, AM_LDFLAGS): Define,
	so that all the usual C compile-and-link machinery comes into play.
	* tests/get-mb-cur-max.c: Include "progname.h".
	Remove unnecessary inclusion of <ctype.h>.
	Mike Frysinger reported the "implicit decl of set_program_name" warning.

	build: detect PCRE support also when <pcre/pcre.h> is the header
	* m4/pcre.m4: Also check for <pcre/pcre.h>.
	* src/pcresearch.c: Include <pcre/pcre.h>, if needed.
	Guard inclusions with HAVE_PCRE_H and HAVE_PCRE_PCRE_H, not HAVE_LIBPCRE.
	* NEWS (Bug fixes): Mention it.
	Dmitry V. Levin reported that PCRE support was not detected
	on systems with <pcre.h> not in the default include path.

2010-03-29  Jim Meyering  <meyering@redhat.com>

	post-release administrivia
	* NEWS: Add header line for next release.
	* .prev-version: Record previous version.
	* cfg.mk (old_NEWS_hash): Auto-update.

	version 2.6.2
	* NEWS: Record release date.

2010-03-29  Eric Blake  <eblake@redhat.com>

	build: avoid warnings on cygwin
	* lib/savedir.c (isdir): Avoid shadowing a declaration.
	* src/main.c (get_nondigit_option): Cast away const to avoid
	compiler warning.

	maint: ignore new test executable
	* .gitignore: Enhance.

2010-03-29  Jim Meyering  <meyering@redhat.com>

	doc: consolidate redundant-looking entries
	* NEWS: Consolidate the two --include/exclude-related entries.
	Suggested by Eric Blake.

2010-03-29  Paolo Bonzini  <bonzini@gnu.org>

	tests: use $(...) consistently
	* tests/backref.sh: Use `...' instead of ``...'' in comments.

Linux-PAM-1.1.2/modules/pam_rootok/pam_rootok.c
/* pam_rootok module */

/*
 * $Id: pam_rootok.c,v 1.8 2010/04/06 08:07:11 kukuk Exp $
 *
 * Written by Andrew Morgan <morgan@linux.kernel.org> 1996/3/11
 */

#include "config.h"

#include <stdio.h>
#include <unistd.h>
#include <syslog.h>
#include <stdarg.h>
#include <string.h>

/*
 * here, we make a definition for the externally accessible function
 * in this file (this definition is required for static a module
 * but strongly encouraged generally) it is used to instruct the
 * modules include file to define the function prototypes.
 */

#define PAM_SM_AUTH

#include <security/pam_modules.h>
#include <security/pam_ext.h>

#ifdef WITH_SELINUX
#include <selinux/selinux.h>
#include <selinux/av_permissions.h>
#endif

/* argument parsing */

#define PAM_DEBUG_ARG       01

static int
_pam_parse (const pam_handle_t *pamh, int argc, const char **argv)
{
    int ctrl=0;

    /* step through arguments */
    for (ctrl=0; argc-- > 0; ++argv) {

	/* generic options */

	if (!strcmp(*argv,"debug"))
	    ctrl |= PAM_DEBUG_ARG;
	else {
	    pam_syslog(pamh, LOG_ERR, "unknown option: %s", *argv);
	}
    }

    return ctrl;
}

static int
check_for_root (pam_handle_t *pamh, int ctrl)
{
    int retval = PAM_AUTH_ERR;

    if (getuid() == 0)
#ifdef WITH_SELINUX
      if (is_selinux_enabled()<1 || checkPasswdAccess(PASSWD__ROOTOK)==0)
#endif
	retval = PAM_SUCCESS;

    if (ctrl & PAM_DEBUG_ARG) {
       pam_syslog(pamh, LOG_DEBUG, "root check %s",
	          (retval==PAM_SUCCESS) ? "succeeded" : "failed");
    }

    return retval;
}

/* --- management functions --- */

PAM_EXTERN int
pam_sm_authenticate (pam_handle_t *pamh, int flags UNUSED,
		     int argc, const char **argv)
{
    int ctrl;

    ctrl = _pam_parse(pamh, argc, argv);

    return check_for_root (pamh, ctrl);
}

PAM_EXTERN int
pam_sm_setcred (pam_handle_t *pamh UNUSED, int flags UNUSED,
		int argc UNUSED, const char **argv UNUSED)
{
    return PAM_SUCCESS;
}

PAM_EXTERN int
pam_sm_acct_mgmt (pam_handle_t *pamh, int flags UNUSED,
		  int argc, const char **argv)
{
    int ctrl;

    ctrl = _pam_parse(pamh, argc, argv);

    return check_for_root (pamh, ctrl);
}

PAM_EXTERN int
pam_sm_chauthtok (pam_handle_t *pamh, int flags UNUSED,
		  int argc, const char **argv)
{
    int ctrl;

    ctrl = _pam_parse(pamh, argc, argv);

    return check_for_root (pamh, ctrl);
}

#ifdef PAM_STATIC

/* static module data */

struct pam_module _pam_rootok_modstruct = {
    "pam_rootok",
    pam_sm_authenticate,
    pam_sm_setcred,
    pam_sm_acct_mgmt,
    NULL,
    NULL,
    pam_sm_chauthtok,
};

#endif

/* end of module definition */
